tests/components/crownstone/test_config_flow.py

```diff
@@ -0,0 +1,531 @@
+    MANUAL_PATH,
+)
+from homeassistant.const import CONF_EMAIL, CONF_PASSWORD
+from homeassistant.core import HomeAssistant
+
+from tests.common import MockConfigEntry
+
+
+@pytest.fixture(name="crownstone_setup", autouse=True)
+def crownstone_setup():
+    """Mock Crownstone entry setup."""
+    with patch(
+        "homeassistant.components.crownstone.async_setup_entry", return_value=True
+    ):
+        yield
+
+
+def get_mocked_crownstone_cloud(spheres: dict[str, MagicMock] | None = None):
+    """Return a mocked Crownstone Cloud instance."""
+    mock_cloud = MagicMock()
+    mock_cloud.async_initialize = AsyncMock()
+    mock_cloud.cloud_data = Spheres(MagicMock(), "account_id")
+    mock_cloud.cloud_data.spheres = spheres
+
+    return mock_cloud
+
+
+def create_mocked_spheres(amount: int) -> dict[str, MagicMock]:
+    """Return a dict with mocked sphere instances."""
+    spheres: dict[str, MagicMock] = {}
+    for i in range(amount):
+        spheres[f"sphere_id_{i}"] = MagicMock()
+        spheres[f"sphere_id_{i}"].name = f"sphere_name_{i}"
+        spheres[f"sphere_id_{i}"].cloud_id = f"sphere_id_{i}"
+
+    return spheres
+
+
+def get_mocked_com_port():
+    """Mock of a serial port."""
+    port = ListPortInfo("/dev/ttyUSB1234")
+    port.device = "/dev/ttyUSB1234"
+    port.serial_number = "1234567"
+    port.manufacturer = "crownstone"
+    port.description = "crownstone dongle - crownstone dongle"
+    port.vid = 1234
+    port.pid = 5678
+
+    return port
+
+
+def create_mocked_entry_data_conf(email: str, password: str):
+    """Set a result for the entry data for comparison."""
+    mock_data: dict[str, str | None] = {}
+    mock_data[CONF_EMAIL] = email
+    mock_data[CONF_PASSWORD] = password
+
+    return mock_data
+
+
+def create_mocked_entry_options_conf(usb_path: str | None, usb_sphere: str | None):
+    """Set a result for the entry options for comparison."""
+    mock_options: dict[str, str | None] = {}
+    mock_options[CONF_USB_PATH] = usb_path
+    mock_options[CONF_USB_SPHERE] = usb_sphere
+
+    return mock_options
+
+
+async def start_config_flow(hass: HomeAssistant, mocked_cloud: MagicMock):
+    """Patch Crownstone Cloud and start the flow."""
+    mocked_login_input = {
+        CONF_EMAIL: "example@homeassistant.com",
+        CONF_PASSWORD: "homeassistantisawesome",
+    }
+
+    with patch(
+        "homeassistant.components.crownstone.config_flow.CrownstoneCloud",
+        return_value=mocked_cloud,
+    ):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": "user"}, data=mocked_login_input
+        )
+
+    return result
+
+
+async def start_options_flow(
+    hass: HomeAssistant, entry_id: str, mocked_cloud: MagicMock
+):
+    """Patch CrownstoneEntryManager and start the flow."""
+    mocked_manager = MagicMock()
+    mocked_manager.cloud = mocked_cloud
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry_id] = mocked_manager
```

Comment:
Please set up the integration instead while patching the library appropriately to avoid I/O. We don't want to access integration details like hass.data in the tests.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations